network/src/counters.rs

```diff
@@ -66,29 +67,91 @@ pub static LIBRA_NETWORK_RPC_MESSAGES: Lazy<IntCounterVec> = Lazy::new(|| {
     register_int_counter_vec!(
         "libra_network_rpc_messages",
         "Number of RPC messages",
-        &["type", "state"]
+        &[
+            "role_type",
+            "network_id",
+            "self_peer_id",
+            "protocol_id",
+            "type",
+            "state"
+        ]
     )
     .unwrap()
 });
 
+pub fn rpc_messages(
+    network_context: &NetworkContext,
+    protocol_id: ProtocolId,
+    type_label: &'static str,
+    state_label: &'static str,
+) -> IntCounter {
+    LIBRA_NETWORK_RPC_MESSAGES.with_label_values(&[
+        network_context.role().as_str(),
+        network_context.network_id().as_cow_str().as_ref(),
+        network_context.peer_id_short_str(),
+        protocol_id.as_str(),
+        type_label,
+        state_label,
+    ])
+}
+
+// TODO(philiphayes): the default histogram buckets don't make sense for measuring
+// message sizes; we need to set these explicitly to something reasonable...
```

Comment:
Here's a good callout.

Right now you can only look at 1 minute buckets, and we should really be looking at them in aggregate.  Do these make sense as a metric?  Does knowing how large a message in aggregate tell us anything?  Otherwise, structured logging might be a better place for per request metrics like this.

Reply:
i wrote this down here because the default histogram buckets are targeted towards measuring "service latency" (https://docs.rs/prometheus/0.10.0/src/prometheus/histogram.rs.html#25) and so they look like 

```rust
pub const DEFAULT_BUCKETS: &[f64; 11] = &[
    0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1.0, 2.5, 5.0, 10.0,
];
```

which doesn't make sense for measuring request sizes, since 99% will end up in the last bucket

as for how useful measuring the rpc message sizes is... I've never had any particular reason to look at them yet, but it might be useful for seeing bandwidth usage by network type and direction. maybe we just put a blanket metric on the actual socket r/w and remove the per-protocol counters?

Reply:
If we're going to log message sizes, I'd say the most important is peer > message type (consensus, mempool) > protocol type (request, reply, send). In general, I don't think we'll get much out of protocol type. Could remove that to reduce the amount of metrics being created.

Reply:
also this might be fine to just be a counter, not sure we get much out of a histogram... I think in ODS we could take something like the following

(value at time t1 - value at time t0) / (t0 - t1) and do that at second or 60 second intervals to make some pretty graphs, not sure what prometheus / grafana offer.